### ISSUE TITLE
Evaluation cache conflates trainset and valset ids: valset scores are silently served from trainset rollouts

### ISSUE THREAD

**Issue by blerchin** [gepa-ai/gepa#424]
## Summary

When `valset` is passed as a separate dataset (the documented **Generalization** mode), GEPA can record a valset score for an example it never evaluated, reusing the rollout of a **different** example from the trainset.

`ListDataLoader` ids are list positions, the trainset and the valset get independent loaders each numbering from `0`, and `EvaluationCache` is keyed `(candidate_hash, example_id)` — so trainset item *i* and valset item *i* collide on one cache key.

This does not affect the modes where `valset=None`: there `val_loader is train_loader`, so the shared id space is correct and the cache does exactly what it should. It only bites in the mode whose purpose is measuring generalization to held-out data.

## Reproduction

```python
"""GEPA: valset scores are served from trainset rollouts when a separate valset is passed.

Note the mock examples deliberately carry no "id" key: GEPA's example ids here are purely
list positions. The "train-N"/"val-N" strings are the example *contents*, named after their
position only so the collision is legible in the output.
"""
import collections
import json

from gepa.core.state import EvaluationCache

TRAIN = [{"content": f"train-{n}"} for n in range(12)]
VAL = [{"content": f"val-{n}"} for n in range(6)]
SLOT_OF = {example["content"]: slot for slot, example in enumerate(VAL)}

# candidate -> valset slot -> content of the example that actually produced that slot's score
produced_by = collections.defaultdict(dict)


def tag_of(candidate):
    raw = next(iter(candidate.values()))
    try:
        return json.loads(raw)["text"]
    except Exception:
        return str(raw)


def ran_content(cached_output):
    """The evaluator records the example it ran; find that marker in the cached output."""
    parts = cached_output if isinstance(cached_output, tuple) else (cached_output,)
    return next((part["ran"] for part in parts if isinstance(part, dict) and "ran" in part), "?")


_get_batch = EvaluationCache.get_batch


def get_batch(self, candidate, example_ids):
    """Valset lookups come through here; a hit fills a slot without running anything."""
    cached, uncached = _get_batch(self, candidate, example_ids)
    for slot, entry in cached.items():
        produced_by[tag_of(candidate)][slot] = ran_content(entry.output)
    return cached, uncached


EvaluationCache.get_batch = get_batch

from gepa.optimize_anything import EngineConfig, GEPAConfig, ReflectionConfig, optimize_anything


def evaluator(candidate, example):
    tag = json.loads(candidate)["text"]
    content = example["content"]
    if content in SLOT_OF:
        produced_by[tag][SLOT_OF[content]] = content
    # monotonically improving candidates, so children are accepted and trigger valset evaluations
    score = 0.1 * (int(tag.split("-")[-1]) + 1 if tag != "seed" else 0)
    return score, {"ran": content}


counter = iter(range(999))
optimize_anything(
    seed_candidate=json.dumps({"text": "seed"}),
    evaluator=evaluator,
    dataset=TRAIN,
    valset=VAL,
    objective="repro",
    config=GEPAConfig(
        engine=EngineConfig(
            max_metric_calls=200,
            max_candidate_proposals=5,
            display_progress_bar=False,
            cache_evaluation=True,
        ),
        reflection=ReflectionConfig(
            reflection_lm=lambda _prompt: json.dumps({"text": f"cand-{next(counter)}"}),
            reflection_minibatch_size=4,
        ),
    ),
)

print("\nContent of the example that actually produced each valset slot's score")
print("(* = a trainset example's rollout, reused for a valset slot)\n")
print("  " + "candidate".ljust(10) + "".join(f"slot {slot}".ljust(11) for slot in range(len(VAL))))
for tag, slots in produced_by.items():
    cells = []
    for slot in range(len(VAL)):
        content = slots.get(slot, "-")
        cells.append((content + ("*" if content.startswith("train-") else "")).ljust(11))
    reused = sum(1 for value in slots.values() if value.startswith("train-"))
    print("  " + tag.ljust(10) + "".join(cells) + (f"  {reused}/{len(VAL)} reused" if reused else ""))
```

Output:

```
Content of the example that actually produced each valset slot's score
(* = a trainset example's rollout, reused for a valset slot)

  candidate slot 0     slot 1     slot 2     slot 3     slot 4     slot 5     
  seed      val-0      val-1      val-2      val-3      val-4      val-5      
  cand-0    val-0      train-1*   val-2      val-3      val-4      train-5*     2/6 reused
  cand-1    val-0      val-1      train-2*   train-3*   val-4      val-5        2/6 reused
  cand-2    train-0*   val-1      val-2      val-3      train-4*   val-5        2/6 reused
  cand-3    val-0      val-1      val-2      train-3*   val-4      val-5        1/6 reused
  cand-4    train-0*   val-1      val-2      val-3      val-4      train-5*     2/6 reused
```

The seed is clean: it is evaluated on the valset before any minibatch rollouts exist to collide with. Every candidate after it has valset slots filled by trainset rollouts, and the collision is plainly positional — valset slot 1 is filled by the rollout of trainset position 1, slot 5 by trainset position 5, and so on. Each of those candidates still ends up with a full six-value valset score vector, and nothing in the run output indicates that two of the six were never evaluated.

## Root cause

1. `ListDataLoader.all_ids()` returns `range(len(items))`, so ids are list positions — `src/gepa/core/data_loader.py:56-57`.
2. Two independent loaders are built when a valset is supplied — `src/gepa/gepa_launcher.py:1381-1382`:
   ```python
   train_loader = ensure_loader(effective_dataset)
   val_loader = ensure_loader(valset) if valset is not None else train_loader
   ```
3. A single `EvaluationCache` serves both, keyed `(candidate_hash, example_id)` — `src/gepa/core/state.py:30,72-87`.
4. The valset path reads the cache (`src/gepa/core/engine.py:306`) while the minibatch path only writes it (`src/gepa/proposer/reflective_mutation/reflective_mutation.py:386,650`). The contamination is therefore one-directional: valset results are replaced by trainset results, never the reverse.

Expected collisions per candidate are about `minibatch_size × |valset| / |trainset|`, so this grows with minibatch size and with the valset fraction.

## Why it matters

The fabricated per-instance scores do not just distort the reported aggregate — they feed candidate selection. `prog_candidate_val_subscores`, `pareto_front_valset` and `program_at_pareto_front_valset` all store the cached value, and `select_program_candidate_from_pareto_front` (`src/gepa/gepa_utils.py:107`) samples parents proportionally to the number of valset instances a candidate holds on the frontier. Each contaminated instance is an extra selection ticket.

The bias is systematically upward, for two compounding reasons: the reused value is the candidate's *own* minibatch result, i.e. the data the reflection step just optimized against, and a child only reaches the valset stage if those minibatch scores were good. So it is in-sample signal, filtered for high scores, injected into the held-out measurement.

In one of our runs (68 train / 18 val, minibatch 10), 11 of 108 valset slot evaluations were substituted this way. Their mean score was 0.895 against 0.659 for the genuinely-run valset evaluations; one candidate's reported valset average was inflated by +0.105 (0.592 → 0.697); and 10 of the 11 placed their candidate on the Pareto front for an instance it never ran, including one instance where the candidate was the *sole* frontier occupant purely on the strength of a different example's rollout.

## Suggested fix

Disambiguate the cache key when the two loaders are not the same object, e.g. namespace by split or by loader identity:

```python
self._cache[(_candidate_hash(candidate), split_key, example_id)] = ...
```

The newer `oa` eval-server path already does the equivalent, defaulting ids to `f"{split_name}_{i}"` (`src/gepa/oa/eval_server.py:205-215`); the core loader/cache layer just hasn't caught up. Alternatives: give `ListDataLoader` an id prefix/offset, or have `ensure_loader` assign each loader it creates a distinct namespace.

An assertion that `val_loader is train_loader` or that their id sets are disjoint would have caught this at startup.

## Workarounds

- `cache_evaluation=False` avoids it completely, at the cost of re-running evaluations — expensive when a single evaluation takes minutes.
- On the released `0.1.4` API, passing a custom `DataLoader` whose `all_ids()` returns stable unique ids (we use dataset-level case ids) fixes it: `ensure_loader` duck-types it through and the valset state is all `dict[DataId, ...]`. On `main` this no longer works through `optimize_anything`, because `EvalServer.__init__` iterates the dataset directly (`src/gepa/oa/eval_server.py:211`) and a loader is not iterable.

## Environment

- Reproduced on `main` (`pip install git+https://github.com/gepa-ai/gepa@main`) and on released `0.1.4`.
- Python 3.10.
